### DISCUSSION TITLE
How do I proxy to a downstream Kubernetes service?

### DISCUSSION THREAD

**Discussion by mschaefer-gresham** [dotnet/yarp#1907] [General] [answered]
Using Ocelot I can specify that my downstream route is a Kubernetes service. I can provice the service name and namepace in my route configuration. Note, this requires adding some K8s specific middleware to Ocelot. 

How can I achieve this with YARP? I do not see any examples.

**Comment by karelz** [comment]
@mschaefer-gresham what exactly do you expect from the Kubernetes service? Anything special beyond just referencing the service?

**Comment by mschaefer-gresham** [comment, marked as answer]
We currently have an api gateway microservice that forwards requests to our downstream services inside the k8s cluster (using Ocelot). You can read about it [here](https://ocelot.readthedocs.io/en/latest/features/kubernetes.html). It would be nice to have comparable functionality in YARP.

**Reply by MihaZupan** [reply]
You should be able to rely on [k8s's DNS](https://kubernetes.io/docs/concepts/services-networking/service/#dns) to resolve the service names for you.
That is, you can configure YARP as you would normally, and use the service name as the hostname for the destination.

```json
"destination1": {
  "Address": "http://my-service/"
}
```

**Reply by adityamandaleeka** [reply]
@mschaefer-gresham Let us know if you have any other issues after you try what @MihaZupan suggested. Marking this as answered for now.

**Reply by Joren-Thijs-KasparSolutions** [reply]
@adityamandaleeka I have a followup question. Since the post talks about comparable functionality to Ocelot. The Ocelot middelware specified will not only discover the addresses of the named kubernetes services (ex. my-service). But it will also automatically load balance between them if multiple instances are found. Can Yarp do something similar in this scenario?

Since in the code sample above, you configure the kubernetes service name in the address property of the destination. But to do load balancing in yarp you need to specify multiple destinations in a cluster...

**Reply by karelz** [reply]
@Joren-Thijs-KasparSolutions can you please file a new question? This is not directly related to the original one. The original one is marked answered, and we do not monitor answered questions anymore. Thanks!

**Comment by mschaefer-gresham** [comment]
The answer did solve my issue. Thank you.
